src/jit/importer.cpp

```diff
@@ -2773,57 +2783,177 @@ GenTreePtr      Compiler::impInitializeArrayIntrinsic(CORINFO_SIG_INFO * sig)
 #endif
         )
     {
-        //
-        // In order to simplify the code, we won't support the multi-dim
-        // case.  Instead we simply return NULL, and the caller will insert 
-        // a run-time call to the helper.  Note that we can't assert
-        // failure here, because this is a valid case.
-        //
+        if (newArrayCall->gtCall.gtCallMethHnd != eeFindHelper(CORINFO_HELP_NEW_MDARR))
+            return nullptr;
 
-        return NULL;
+        isMDArray = true;
     }
 
+    CORINFO_CLASS_HANDLE arrayClsHnd = (CORINFO_CLASS_HANDLE) newArrayCall->gtCall.compileTimeHelperArgumentHandle;
 
     //
-    // Make sure there are exactly two arguments:  the array class and
-    // the number of elements.
+    // Make sure we found a compile time handle to the array
     //
 
-    GenTreePtr arrayLengthNode;
+    if (!arrayClsHnd)
+        return nullptr;
+
+    unsigned rank;
+    S_UINT32 numElements;
 
-    GenTreeArgList* args = newArrayCall->gtCall.gtCallArgs;
-#ifdef FEATURE_READYTORUN_COMPILER
-    if (newArrayCall->gtCall.gtCallMethHnd == eeFindHelper(CORINFO_HELP_READYTORUN_NEWARR_1))
+    if (isMDArray)
     {
-        // Array length is 1st argument for readytorun helper
-        arrayLengthNode = args->Current();
+        rank = info.compCompHnd->getArrayRank(arrayClsHnd);
+
+        if (rank == 0)
+            return nullptr;
+
+        GenTreeArgList* args = newArrayCall->gtCall.gtCallArgs;
+        GenTreeArgList* numArgsArg = nullptr;
+        GenTreeArgList* beginArgs = nullptr;
+        GenTreeArgList* endArgs = nullptr;
+
+        if (Target::g_tgtArgOrder == Target::ARG_ORDER_R2L)
+        {
+            numArgsArg = args->Rest();
+
+            if (numArgsArg != nullptr)
+                beginArgs = numArgsArg->Rest();
+        }
+        else
+        {
+            beginArgs = args;
+
+            for (GenTreeArgList* arg = args; arg != nullptr; arg = arg->Rest())
+            {
+                bool isSecondToLast = (arg->Rest() != nullptr) && (arg->Rest()->Rest() == nullptr);
+
+                if (isSecondToLast)
+                {
+                    numArgsArg = arg;
+                    endArgs = arg;
+                    break;
+                }
+            }
+        }
+
+        //
+        // The number of arguments should be a constant between 1 and 64. The rank can't be 0
+        // so at least one length must be present and the rank can't exceed 32 so there can
+        // be at most 64 arguments - 32 lengths and 32 lower bounds.
+        //
+
+        if ((beginArgs == nullptr) ||
+            (numArgsArg == nullptr) ||
+            (numArgsArg->Current()->OperGet() != GT_CNS_INT) ||
```

Comment:
Above condition could be written using IsCnIntOrI()

!numArgsArg->Current()->IsCnsIntOrI()